src/Semver.elm

```diff
@@ -213,63 +213,123 @@ parser =
         |= nat
         |. Parser.symbol "."
         |= nat
-        |= Parser.oneOf [ preReleaseVersion, Parser.succeed [] ]
-        |= Parser.oneOf [ buildMetadata, Parser.succeed [] ]
+        |= Parser.oneOf
+            [ identifierSeries preReleaseIdentifier (Parser.symbol "-")
+            , Parser.succeed []
+            ]
+        |= Parser.oneOf
+            [ identifierSeries buildMetadataIdentifier (Parser.symbol "+")
+            , Parser.succeed []
+            ]
         |. Parser.end
 
 
 {-| Parses only natural numbers (including zero).
 -}
 nat : Parser Int
 nat =
-    let
-        isBetween low high char =
-            let
-                code =
-                    Char.toCode char
-            in
-                (code >= Char.toCode low) && (code <= Char.toCode high)
-    in
-        Parser.succeed identity
-            |= Parser.oneOf
-                [ Parser.keep (Parser.Exactly 1) ((==) '0')
-                , Parser.succeed (++)
-                    |= Parser.keep (Parser.Exactly 1) (isBetween '1' '9')
-                    |= Parser.keep zeroOrMore Char.isDigit
-                ]
-            |> Parser.andThen
-                (\natStr ->
-                    case String.toInt natStr of
-                        Ok nat ->
-                            Parser.succeed nat
-
-                        Err err ->
-                            Parser.fail err
-                )
-
-
-{-| An alphanumeric identifier, [a-zA-Z0-9-]+.
--}
-identifier : Parser String
-identifier =
-    Parser.keep oneOrMore (\c -> c == '-' || Char.isLower c || Char.isUpper c || Char.isDigit c)
+    Parser.succeed identity
+        |= numericIdentifier
+        |> Parser.andThen
+            (\natStr ->
+                case String.toInt natStr of
+                    Ok nat ->
+                        Parser.succeed nat
+
+                    Err err ->
+                        Parser.fail err
+            )
 
 
-preReleaseVersion : Parser (List String)
-preReleaseVersion =
-    identifierSeries (Parser.symbol "-")
+preReleaseIdentifier : Parser String
+preReleaseIdentifier =
+    Parser.oneOf [ alphanumericIdentifier, numericIdentifier ]
 
 
-buildMetadata : Parser (List String)
-buildMetadata =
-    identifierSeries (Parser.symbol "+")
+buildMetadataIdentifier : Parser String
+buildMetadataIdentifier =
+    Parser.oneOf [ alphanumericIdentifier, digits ]
 
 
-identifierSeries : Parser a -> Parser (List String)
-identifierSeries seperator =
+identifierSeries : Parser i -> Parser a -> Parser (List i)
+identifierSeries identifier seperator =
     Parser.succeed (::)
         |. seperator
         |= identifier
         |= Parser.repeat
             zeroOrMore
             (Parser.succeed identity |. Parser.symbol "." |= identifier)
+
+
+{-| An alphanumeric identifier, [a-zA-Z0-9-]+, containing at least one non-digit.
+-}
+alphanumericIdentifier : Parser String
+alphanumericIdentifier =
```

Comment:
Did you consider writing it as follow ?

```
- zeroOrMore isDigit
- one nonDigit
- zeroOrMore isIdentifierCharacter
```

Reply:
You mean to compose `alphanumericIdentifier` in this way? This will not work because it will make the parser commit too early.

Let's say there are digits in the beginning. Then on seeing those, the parser will commit. Let's say there are no digits in the beginning. Then on seeing those, the parser will _also_ commit (if memory serves me right), because it will succeed with zero chars read.

The way it is written is the closest late committing parser that I could think of:

- Either something starting in a non-digit (commit right away)
- Or a digit (don't commit yet), followed by a non-digit (commit) and some other stuff

Reply:
With the use of `Parser.delayedCommit`, I think the following parser would work:

`(Parser.delayedCommit (zeroOrMore isDigit) (one nonDigit)) `followedBy` (zeroOrMore isIdentifierCharacter)`

According to the [doc](https://old.elm.dmy.fr/packages/elm-tools/parser/2.0.1/Parser#delayedCommit):

> Only commit if `Parser a` succeeds and `Parser value` makes some progress.

That means `nonDigit` made some progress, that means it’s certain it’s alphanumeric.

Reply:
You need `delayedCommitMap` because you want both parts. I think the total amount of `(++)` operations would be 2 in both cases. It might allow to shed the `oneOf`, but seems like code golfing to me. Do you find one version more comprehensible?

Reply:
I think it would become simpler to read, because we would get rid of the `oneOf`. But I’m ok to keep the current version.